ouroboros-consensus/src/Ouroboros/Consensus/Ledger/SupportsMempool.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeFamilies     #-}
+module Ouroboros.Consensus.Ledger.SupportsMempool (
+    LedgerSupportsMempool (..)
+  , HasTxId (..)
+  , GenTxId
+  , HasTxs (..)
+  ) where
+
+import           Control.Monad.Except
+import           Data.Word (Word32)
+import           GHC.Stack (HasCallStack)
+
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Util.IOLike
+
+
+class ( UpdateLedger blk
+      , NoUnexpectedThunks (GenTx blk)
+      , Show (GenTx blk)
+      , Show (ApplyTxErr blk)
+      ) => LedgerSupportsMempool blk where
+  -- | Generalized transaction
+  --
+  -- The mempool (and, accordingly, blocks) consist of "generalized
+  -- transactions"; this could be "proper" transactions (transferring funds) but
+  -- also other kinds of things such as update proposals, delegations, etc.
+  data family GenTx blk :: *
+
+  -- | Check whether the internal invariants of the transaction hold.
+  txInvariant :: GenTx blk -> Bool
+  txInvariant = const True
+
+  -- | Updating the ledger with a single transaction may result in a different
+  -- error type as when updating it with a block
+  type family ApplyTxErr blk :: *
+
+  -- | Apply transaction we have not previously seen before
+  applyTx :: LedgerConfig blk
+          -> GenTx blk
+          -> TickedLedgerState blk
+          -> Except (ApplyTxErr blk) (TickedLedgerState blk)
+
+  -- | Re-apply a transaction
+  --
+  -- When we re-apply a transaction to a potentially different ledger state
+  -- expensive checks such as cryptographic hashes can be skipped, but other
+  -- checks (such as checking for double spending) must still be done.
+  reapplyTx :: HasCallStack
+            => LedgerConfig blk
+            -> GenTx blk
+            -> TickedLedgerState blk
+            -> Except (ApplyTxErr blk) (TickedLedgerState blk)
+
+  -- | The maximum number of bytes worth of transactions that can be put into
+  -- a block according to the currently adopted protocol parameters of the
+  -- ledger state.
+  --
+  -- This is (conservatively) computed by subtracting the header size and any
+  -- other fixed overheads from the maximum block size.
+  maxTxCapacity :: LedgerState blk -> Word32
+
+  -- | The maximum transaction size in bytes according to the currently
+  -- adopted protocol parameters of the ledger state.
+  maxTxSize :: LedgerState blk -> Word32
```

Comment:
TODO replace with `TickedLedgerState`

Reply:
We currently don't use `maxTxSize` anywhere. We provide it for the benefit of downstream users, which use `ChainDB.getCurrentLedger` to get the `(Ext)LedgerState`. Requiring a ticked one is rather annoying for them, I suspect. So we'll leave this as is.